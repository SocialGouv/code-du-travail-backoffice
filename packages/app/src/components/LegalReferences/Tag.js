import PropTypes from "prop-types";
import React from "react";

import * as C from "../../constants";
import Button from "../../elements/Button";
import Icon from "../../elements/Icon";
import SavingSpinner from "../../elements/SavingSpinner";
import cdtnApi from "../../libs/cdtnApi";
import LegalReferenceProps from "../../props/LegalReference";
import { ButtonsContainer, Container, Label, Tooltip } from "./Tag.style";
import TagEditorWithClickOutside from "./TagEditor";

const BASE_URL = {
  agreement: "https://beta.legifrance.gouv.fr/conv_coll/id/",
  labor_code: "https://beta.legifrance.gouv.fr/codes/article_lc/",
};

class Tag extends React.PureComponent {
  constructor(props) {
    super(props);
    const { category, dila_id } = props;

    this.isLegacy = category !== null && dila_id === null;
    this.$input = null;

    this.state = {
      content: "",
      isEditing: false,
      isLoading: true,
      isObsolete: false,
    };

    this.edit = this.edit.bind(this);
    this.open = this.open.bind(this);
    this.unedit = this.unedit.bind(this);
    this.update = this.update.bind(this);
  }

  componentDidMount() {
    this.loadUpdatedData();
  }

  async loadUpdatedData() {
    const { category, dila_cid, dila_id } = this.props;
    if (dila_id === null) {
      this.setState({ isLoading: false });

      return;
    }

    try {
      const path = category === C.LEGAL_REFERENCE_CATEGORY.AGREEMENT ? "/agreement" : "/code";
      const { content, id } = await cdtnApi.get(`${path}/article/${dila_cid}`);

      this.setState({
        content,
        isLoading: false,
        isObsolete: id !== dila_id,
      });
    } catch (err) {
      this.setState({
        isLoading: false,
        isObsolete: true,
      });
    }
  }

  edit() {
    this.setState({ isEditing: true });
  }

  unedit() {
    this.setState({ isEditing: false });
  }

  update(value, url) {
    const { answer_id, id, category, dila_cid, dila_container_id, dila_id, onChange } = this.props;

    const data = {
      answer_id,
      category,
      dila_cid,
      dila_container_id,
      dila_id,
      id,
      url,
      value,
    };

    onChange(data);
  }

  open() {
    const { category, dila_container_id, dila_id, url } = this.props;

    if (url !== null) {
      window.open(url, "_blank");

      return;
    }

    const legifranceUriExtra =
      dila_container_id !== null ? `/?idConteneur=${dila_container_id}` : "";
    const legifranceUri = `${BASE_URL[category]}${dila_id}${legifranceUriExtra}`;

    window.open(legifranceUri, "_blank");
  }

  renderLabel() {
    const { value } = this.props;
    const { isObsolete } = this.state;

    return (
      <Label data-testid="label">
        {isObsolete && <Icon color="red" icon="exclamation-circle" withMarginRight />}
        {value}
      </Label>
    );
  }

  renderButtons() {
    const { category, dila_id, id, isReadOnly, onRemove, url } = this.props;

    const parts = [];

    if (dila_id !== null || url !== null) {
      parts.push(
        <Button
          data-testid="button-open"
          icon="link"
          isSmall
          isTransparent
          key="button-open"
          onClick={this.open}
        />,
      );
    }

    if (!isReadOnly && category !== C.LEGAL_REFERENCE_CATEGORY.LABOR_CODE) {
      parts.push(
        <Button
          data-testid="button-edit"
          icon="pen"
          isSmall
          isTransparent
          key="button-edit"
          onClick={this.edit}
        />,
      );
    }

    if (!isReadOnly) {
      parts.push(
        <Button
          data-testid="button-remove"
          icon="trash"
          isSmall
          isTransparent
          key="button-remove"
          onClick={() => onRemove(id)}
        />,
      );
    }

    return <ButtonsContainer>{parts}</ButtonsContainer>;
  }

  render() {
    const { isLoading } = this.state;

    if (isLoading) {
      return (
        <Container alignItems="center" justifyContent="center">
          <SavingSpinner color="var(--color-eerie-black)" size={16} />
        </Container>
      );
    }

    const { id, url, value } = this.props;
    const { content, isEditing } = this.state;
    const hasContent = content.length !== 0;

    if (isEditing) {
      return (
        <Container isEditing>
          <TagEditorWithClickOutside
            data-testid="tag-editor"
            onCancel={this.unedit}
            onSubmit={this.update}
            url={url}
            value={value}
          />
        </Container>
      );
    }

    return (
      <Container
        alignItems="flex-start"
        data-for={id}
        data-tip={content}
        isLegacy={this.isLegacy}
        justifyContent="space-between"
      >
        {this.renderLabel()}
        {this.renderButtons()}
        {hasContent && (
          <Tooltip
            clickable
            data-testid="tooltip"
            delayHide={250}
            effect="solid"
            id={id}
            place="bottom"
            type="light"
          />
        )}
      </Container>
    );
  }
}

Tag.propTypes = {
  ...LegalReferenceProps,
  isReadOnly: PropTypes.bool,
  onChange: PropTypes.func,
  onRemove: PropTypes.func,
};

export default Tag;
