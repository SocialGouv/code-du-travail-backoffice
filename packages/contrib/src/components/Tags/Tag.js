import styled from "@emotion/styled";
import React from "react";
import { Flex } from "rebass";

const Container = styled(Flex)`
  background-color: lightgray;
  border: solid 1px gray;
  border-radius: 0.25rem;
  font-size: 0.875rem;
  margin: 0.5rem 0.5rem 0 0;
  max-width: 32rem;
  padding: 0.2rem 0.4rem;
  user-select: none;
`;
const Text = styled.span`
  overflow: hidden;
  text-overflow: ellipsis;
  white-space: nowrap;
`;
const Button = styled.img`
  cursor: pointer;
  height: 0.75rem;
  margin-left: 0.5rem;
  opacity: 0.5;
  vertical-align: 2px;
  width: 0.75rem;

  :hover {
    opacity: 1;
  }
`;

export default class Tag extends React.PureComponent {
  openUrl(url) {
    window.open(url, "_blank");
  }

  render() {
    const { onRemove, url, value } = this.props;
    const ariaName = this.props.ariaName !== undefined ? this.props.ariaName : "l'Ã©tiquette";

    return (
      <Container alignItems="center">
        <Text>{value}</Text>
        {typeof url === "string" && (
          <Button
            alt={`Bouton ouvrant l'url de ${ariaName} ${value}`}
            onClick={() => this.openUrl(url)}
            src="/static/images/link.svg"
          />
        )}
        <Button
          alt={`Bouton supprimant ${ariaName} ${value}`}
          onClick={() => onRemove(value)}
          src="/static/images/delete.svg"
        />
      </Container>
    );
  }
}
