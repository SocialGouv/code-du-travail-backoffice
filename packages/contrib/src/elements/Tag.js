import styled from "styled-components";

export default styled.span`
  background-color: ${props =>
    props.selected ? "var(--color-lapis-lazuli)" : "transparent"};
  border-radius: 0.8rem;
  border: solid 1px var(--color-lapis-lazuli);
  color: ${props => (props.selected ? "white" : "var(--color-lapis-lazuli)")};
  cursor: pointer;
  font-size: 0.8rem;
  margin: 0 0.5rem 0.5rem 0;
  opacity: ${({ isDisabled }) => (Boolean(isDisabled) ? 0.25 : 1)};
  padding: 0.25rem 0.75rem;
  user-select: none;
  white-space: nowrap;
`;
