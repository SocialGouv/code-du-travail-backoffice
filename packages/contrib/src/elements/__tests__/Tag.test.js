import React from "react";
import { render } from "@testing-library/react";
import Tag from "../Tag";

describe("[Contrib] elements/<Tag />", () => {
  it("should match snapshot", () => {
    const { container } = render(<Tag />);

    expect(container).toMatchSnapshot();
  });
});
