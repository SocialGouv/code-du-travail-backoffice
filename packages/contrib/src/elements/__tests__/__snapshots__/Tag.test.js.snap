// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`[Contrib] elements/<Tag /> should match snapshot 1`] = `
.emotion-0 {
  background-color: transparent;
  border-radius: 0.8rem;
  border: solid 1px var(--color-lapis-lazuli);
  color: var(--color-lapis-lazuli);
  cursor: pointer;
  font-size: 0.8rem;
  margin: 0 0.5rem 0.5rem 0;
  opacity: 1;
  padding: 0.25rem 0.75rem;
  white-space: nowrap;
}

<div>
  <span
    class="emotion-0 emotion-1"
  />
</div>
`;

exports[`[Contrib] elements/<Tag /> should match snapshot with \`isDisabled\` prop 1`] = `
.emotion-0 {
  background-color: transparent;
  border-radius: 0.8rem;
  border: solid 1px var(--color-lapis-lazuli);
  color: var(--color-lapis-lazuli);
  cursor: pointer;
  font-size: 0.8rem;
  margin: 0 0.5rem 0.5rem 0;
  opacity: 0.25;
  padding: 0.25rem 0.75rem;
  white-space: nowrap;
}

<div>
  <span
    class="emotion-0 emotion-1"
  />
</div>
`;
